modules/darwinosl/darwinosl-source.m

```diff
@@ -0,0 +1,559 @@
+    {
+      [self->osLogSource closeAll];
+      return FALSE;
+    }
+
+  if (positionRestored)
+    {
+      OSLogEntry *nextLogEntry = [self->osLogSource fetchNextEntry];
+      if (nextLogEntry)
+        _check_restored_postion(self, nextLogEntry);
+      else
+        msg_warning("darwinosl: Could not fetch log entry after a successful position restore");
+    }
+
+  self->log_source_position.last_used_filter_predicate_hash = (self->options.filter_predicate ?
+                                                              g_str_hash(self->options.filter_predicate) :
+                                                              0);
+  return TRUE;
+}
+
+static void
+_close_osl(LogThreadedSourceDriver *s)
+{
+  DarwinOSLogSourceDriver *self = (DarwinOSLogSourceDriver *)s;
+  [self->osLogSource closeAll];
+}
+
+/* runs in a dedicated thread */
+static LogThreadedFetchResult
+_fetch(LogThreadedSourceDriver *s)
+{
+  DarwinOSLogSourceDriver *self = (DarwinOSLogSourceDriver *)s;
+
+  gboolean fetchedEnough = (self->options.fetch_limit && self->curr_fetch_in_run > self->options.fetch_limit);
+  OSLogEntry *nextLogEntry = (fetchedEnough ? nil :[self->osLogSource fetchNextEntry]);
+
+  if (nextLogEntry == nil)
+    {
+      msg_trace(nextLogEntry ? "darwinosl: Fetch limit reached" : "darwinosl: No more log data currently");
+      msg_debug("darwinosl: Fetched logs in this run", evt_tag_long("fetch count", (long) self->curr_fetch_in_run));
+      LogThreadedFetchResult result = {THREADED_FETCH_NO_DATA, NULL};
+      return result;
+    }
+
+  self->curr_fetch_in_run++;
+
+  const char *log_string = [self->osLogSource stringFromDarwinOSLogEntry:nextLogEntry].UTF8String;
+  msg_trace("darwinosl: Msg string is composed", evt_tag_str("msg", log_string));
+
+  /* NOTE: No other options unfortunately (there is no "current position" like getter)
+            If there are multiple occurrences of the same time, the method returns the earliest occurrence.
+            So, in a worst case there might be log duplications on restart from bookmark
+  */
+  self->log_source_position.log_position = [nextLogEntry.date timeIntervalSince1970];
+  self->log_source_position.last_msg_hash = g_str_hash(log_string);
+
+  LogSource *worker = (LogSource *) self->super.worker;
+  Bookmark *bookmark = ack_tracker_request_bookmark(worker->ack_tracker);
+  darwinosl_source_persist_fill_bookmark(self->log_source_persist, bookmark, self->log_source_position);
+  msg_trace("darwinosl: Bookmark created",
+            evt_tag_printf("position", "%.6f", self->log_source_position.log_position),
+            evt_tag_printf("last message hash", "%u", self->log_source_position.last_msg_hash),
+            evt_tag_printf("filter predicate hash", "%u", self->log_source_position.last_used_filter_predicate_hash));
+
+  LogMessage *msg;
+  gsize msg_len = _log_message_from_string(log_string, self->options.format_options, &msg);
```

Comment:
I was wondering if we need the round-trip to the LogMessage becoming an RFC5424 formatted string and then reparsed into a LogMessage.

Reply:
I guess the other option would be to compose the LogMessage field-by-field manually?

I'm not sure if I could compose it more optimally, but of course, I can give it a try ;)

Reply:
As i can recall i used a LogMessage because 
- it helps keep everything in one container, and allows an easier passing for further processing
- it can parse nicely the string out of the box and gives access to each msg elements when needed
- this fetch is conforming to  _LogThreadedFetcherDriver::LogThreadedFetchResult (*fetch)(LogThreadedFetcherDriver *self), and there was a version when this was part of a real _super.fetch implementation